File: 199-binary-tree-right-side-view/binary-tree-right-side-view.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
    return [] unless root
    queue = []
    level = []
    queue << root
    while queue.size > 0 do
        len = queue.size
        prev = nil
        len.times do
            el = queue.shift
            prev = el.val
            queue << el.left if el.left
            queue << el.right if el.right
        end
        level << prev
    end
    level
end

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        queue = deque()
        queue.append(root)
        ans = []
        while queue:
            length = len(queue)
            ans.append(queue[-1].val)
            while length > 0:
                node = queue.popleft()
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
                length -= 1
        return ans
        
        